android/src/main/java/com/swmansion/gesturehandler/core/GestureHandler.kt

```diff
@@ -688,6 +695,46 @@ open class GestureHandler<ConcreteGestureHandlerT : GestureHandler<ConcreteGestu
   protected open fun onReset() {}
   protected open fun onCancel() {}
 
+  private fun isButtonInConfig(clickedButton: Int): Boolean {
+    if (mouseButton == 0) {
+      return clickedButton == MotionEvent.BUTTON_PRIMARY
+    }
+
+    return clickedButton and mouseButton != 0
+  }
+
+  protected fun shouldActivateWithMouse(sourceEvent: MotionEvent): Boolean {
+    // While using mouse, we get both sets of events, for example ACTION_DOWN and ACTION_BUTTON_PRESS. That's why we want to take actions to only one of them.
+    // On API >= 23, we will use events with infix BUTTON, otherwise we use standard action events (like ACTION_DOWN).
+
+    with(sourceEvent) {
+      // To use actionButton, we need API >= 23.
+      if (getToolType(0) == MotionEvent.TOOL_TYPE_MOUSE && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+        // While using mouse, we want to ignore default events for touch.
+        if (action == MotionEvent.ACTION_DOWN || action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_POINTER_UP || action == MotionEvent.ACTION_POINTER_DOWN) {
+          return@shouldActivateWithMouse false
+        }
+
+        // We don't want to do anything if wrong button was clicked. If we received event for BUTTON, we have to use actionButton to get which one was clicked.
+        if (action != MotionEvent.ACTION_MOVE && !isButtonInConfig(actionButton)) {
+          return@shouldActivateWithMouse false
+        }
+
+        // When we receive ACTION_MOVE, we have to check buttonState field.
+        if (action == MotionEvent.ACTION_MOVE && !isButtonInConfig(buttonState)) {
+          return@shouldActivateWithMouse false
+        }
+      } else if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+        // We do not fully support mouse below API 23, so we will ignore BUTTON events.
+        if (action == MotionEvent.ACTION_BUTTON_PRESS || action == MotionEvent.ACTION_BUTTON_RELEASE) {
+          return@shouldActivateWithMouse false
+        }
+      }
```

Comment:
I think the check for API version can be moved to the top of the method, and we can always return `true` if below `M` - the behavior will stay as it is currently on those platforms, and above API 23 the new logic will run. Additionally, I believe that the second API check (`>= M`) would be unnecessary.

Correct me if I'm wrong though.

Reply:
The main problem is that mouse clicks generate 2 events. This function acts like a filter - it assures that handler will receive only one of them. If we return `true` all the time, handlers will receive both events (like `ACTION_DOWN` and `ACTION_BUTTON_PRESS`).  While I haven't tested if anything breaks in that case, I don't think it's a good idea to leave it that way.